[14.0][ADD] project_sequence

anddago78 opened the pull request:
Add a sequence field to projects, filled automatically and add a code sequence filter in tree view project.

MT-1506 @moduon @rafaelbn @Shide @yajo please review :)

OCA-git-bot committed:
[14.0][ADD] project_sequence

yajo commented on README.md:
```diff
@@ -37,8 +37,8 @@ addon | version | maintainers | summary
-[project_stock](project_stock/) | 14.0.1.1.3 | [![victoralmau](https://github.com/victoralmau.png?size=30px)](https://github.com/victoralmau) | Project Stock
-[project_stock_product_set](project_stock_product_set/) | 14.0.1.0.1 | [![victoralmau](https://github.com/victoralmau.png?size=30px)](https://github.com/victoralmau) | Project Stock Product Set
+[project_stock](project_stock/) | 14.0.1.1.4 | [![victoralmau](https://github.com/victoralmau.png?size=30px)](https://github.com/victoralmau) | Project Stock
+[project_stock_product_set](project_stock_product_set/) | 14.0.1.0.2 | [![victoralmau](https://github.com/victoralmau.png?size=30px)](https://github.com/victoralmau) | Project Stock Product Set
```
It's weird that you have this diff here. This is something usually the bot does for you.

It's also weird that the bot is the author of the commit: 
![imagen](https://user-images.githubusercontent.com/973709/230841694-ec978220-1c92-45ac-95ef-c8f932312925.png)

I think you maybe did something weird with git by accident. Please try this:

1. Undo commit.
2. Remove this diff.
3. Re-do the commit.

yajo commented on project_sequence/data/ir_sequence.xml:
```diff
@@ -0,0 +1,16 @@
+        </record>
+        </odoo>
+
+</data>
```
You can remove `<odoo>` and put `noupdate="1"` directly on `<data>`, and it'll be the same:

```suggestion
<data noupdate="1">

        <record id="seq_worksheet-sequence" model="ir.sequence">
        <field name="name">PROJECT SEQUENCE</field>
        <field name="code">project.sequence</field>
        <field name="prefix">PRJ</field>
        <field name="padding">5</field>
        <field name="company_id" eval="False" />
        </record>

</data>
```

yajo commented on project_sequence/data/ir_sequence.xml:
```diff
@@ -0,0 +1,16 @@
+<data>
+
+    <odoo noupdate="1">
+        <record id="seq_worksheet-sequence" model="ir.sequence">
```
All XML-IDS are lowercase_separated_by_underscores by convention:

```suggestion
        <record id="seq_worksheet_sequence" model="ir.sequence">
```

yajo commented on project_sequence/data/ir_sequence.xml:
```diff
@@ -0,0 +1,16 @@
+
+    <odoo noupdate="1">
+        <record id="seq_worksheet-sequence" model="ir.sequence">
+        <field name="name">PROJECT SEQUENCE</field>
```
DON'T CRY!

```suggestion
        <field name="name">Project sequence</field>
```

yajo commented on project_sequence/i18n/.empty:
This file is here as a placeholder only, because Git would remove directories that contain no files.

However, please:
1. translate to Spanish using POEdit
2. put in this folder the .pot and .po files.
3. remove this `.empty` file.

yajo commented on project_sequence/models/project_project.py:
```diff
@@ -0,0 +1,38 @@
+    _inherit = "project.project"
+
+    code = fields.Char(
+        string="Number sequence",
```
"Code" seems a good name. Remove the string, so Odoo uses the default:

```suggestion
```

yajo commented on project_sequence/tests/test_something.py:
The name of the file "test_something.py" is just an example from the template.

You usually change "something" for... something better. Maybe `test_project_sequence.py`.

yajo commented on project_sequence/tests/test_something.py:
```diff
@@ -0,0 +1,37 @@
+
+
+@freeze_time("2023-01-01 12:00:00", tick=True)
+class SomethingCase(TransactionCase):
```
Here too, "Something" is just a placeholder. Give it a meaningful name:

```suggestion
class ProjectSequenceCase(TransactionCase):
```

yajo commented on project_sequence/readme/INSTALL.rst:
```diff
@@ -0,0 +1,10 @@
+
+To install this module, you need to:
+
+#. Install the module.
```
This is absurd 😆 

As you can see in the comment above, if there are no special installation instructions, just remove the file 🔥

yajo commented on project_sequence/readme/USAGE.rst:
```diff
@@ -0,0 +1,12 @@
+
+To use this module, you need to:
+
+#. Install the module
```
This step is not necessary. It's obvious for any module. Remove it please.

yajo commented on project_sequence/readme/USAGE.rst:
```diff
@@ -0,0 +1,12 @@
+To use this module, you need to:
+
+#. Install the module
+#. Now when create a new project apear a new field with the sequence.
```
This is not a valid usage section. You're describing what the module does, but you're not explaining how a user can make it work.

Don't assume whoever will install the module knows where projects are in Odoo. We'll have a meet about this later.

yajo commented on project_sequence/static/description/icon.svg:
Please remove this file.

The SVG is provided to make editions easier. However, when you're just using the default icon, you don't need to include it.

yajo commented on project_sequence/views/project_project.xml:
```diff
@@ -0,0 +1,38 @@
+    <field name="arch" type="xml">
+        <xpath expr="//field[@name='partner_id']" position="after">
+            <field name="code" />
+            </xpath>
```
Be consistent with indentation, to make code more readable:

```suggestion
        </xpath>
```

yajo commented on project_sequence/views/project_project.xml:
```diff
@@ -0,0 +1,38 @@
+            <xpath expr="//field[@name='partner_id']" position="after">
+                <field name="display_name" />
+                <field name="code" optional="hide" />
+                </xpath>
+
```
```suggestion
            </xpath>
```

yajo commented on project_sequence/views/project_project.xml:
```diff
@@ -0,0 +1,38 @@
+    <field name="inherit_id" ref="project.view_project_project_filter" />
+    <field name="arch" type="xml">
+        <xpath expr="//field[@name='partner_id']" position="after">
+                <field name="code" />
```
Use consistent indentation please:

```suggestion
            <field name="code" />
```

BTW, it was expected that the *default* search included the code. If the default search is using `display_name`, that'll be the case automatically (because the code's always there).

yajo commented on project_sequence/views/project_project.xml:
```diff
@@ -0,0 +1,38 @@
+    <field name="model">project.project</field>
+    <field name="inherit_id" ref="project.edit_project" />
+    <field name="arch" type="xml">
+        <xpath expr="//field[@name='partner_id']" position="after">
```
Please pay attention to the module requirements. The field goes into the "settings" tab.

anddago78 replied:
Hi, the partner_id is in settings tab.

yajo commented on project_sequence/models/project_project.py:
```diff
@@ -0,0 +1,38 @@
+    name = fields.Char(required=False)
+    display_name = fields.Char(compute="_compute_display_name_project")
+
+    def _compute_display_name_project(self):
```
Remember to add dependencies:

```suggestion
    @api.depends("name", "code")
    def _compute_display_name_project(self):
```

yajo commented on project_sequence/models/project_project.py:
```diff
@@ -0,0 +1,38 @@
+        default=lambda self: self.env["ir.sequence"].next_by_code("project.sequence"),
+    )
+    name = fields.Char(required=False)
+    display_name = fields.Char(compute="_compute_display_name_project")
```
You can make it stored and indexed, so users can search on the display name directly:

```suggestion
    display_name = fields.Char(compute="_compute_display_name_project", store=True, index=True)
```

yajo reviewed (changes requested):
It's OK in general. It just has a lot of little details to improve. See suggestions below. Thanks!

fcvalgar reviewed (commented):
After my functional review I detect that each project creation generates a new sequence.

I think what you wanted to do @anddago78 is to set a single sequence to project creation.

Can you check if this point is correct. Thank you.

![Captura desde 2023-04-10 12-41-03](https://user-images.githubusercontent.com/17847432/230888609-81c16355-24a7-4d47-9f0a-4f6f20c4f5a8.png)

yajo commented on project_sequence/data/ir_sequence.xml:
```diff
@@ -0,0 +1,12 @@
+    <record id="seq_project_sequence" model="ir.sequence">
+        <field name="name">Project sequence</field>
+        <field name="code">project.sequence</field>
+        <field name="prefix">%(y)s-</field>
```
If you use a year prefix, I think this will go in 2024 from `23-00111` to `24-00112`.

You should probably add time-based automatic subsequences to make it go from `23-00111` to `24-00001`.
Could you double-check this please?

yajo commented on project_sequence/i18n/es.po:
```diff
@@ -0,0 +1,48 @@
+#. module: project_sequence
+#: model:ir.model.fields,field_description:project_sequence.field_project_project__code
+msgid "Code"
+msgstr "Codigo"
```
```suggestion
msgstr "Código"
```

yajo commented on project_sequence/i18n/es.po:
```diff
@@ -0,0 +1,48 @@
+#. module: project_sequence
+#: model:ir.model.fields,field_description:project_sequence.field_project_project__id
+msgid "ID"
+msgstr "Id"
```
```suggestion
msgstr "ID"
```

yajo commented on project_sequence/i18n/es.po:
```diff
@@ -0,0 +1,48 @@
+#. module: project_sequence
+#: model:ir.model.fields,field_description:project_sequence.field_project_project____last_update
+msgid "Last Modified on"
+msgstr "Ultima modificación"
```
```suggestion
msgstr "Última modificación el"
```

yajo commented on project_sequence/models/project_project.py:
```diff
@@ -0,0 +1,45 @@
+class ProjectProject(models.Model):
+    _inherit = "project.project"
+
+    codesequence = fields.Char(
```
Let's give this field a more natural name.

```suggestion
    sequence_code = fields.Char(
```

yajo commented on project_sequence/models/project_project.py:
```diff
@@ -0,0 +1,45 @@
+        required=False,
+    )
+    display_name = fields.Char(
+        compute="_compute_display_name_project", store=True, index=True
```
You can see that [`display_name` has a compute method enabled by default](https://github.com/odoo/odoo/blob/680ff54c56e3a1c1525999c448385417aa7dbb0f/odoo/models.py#L1715-L1730). You can see there that the method depends on the results of the `name_get` function.

So you don't need to override the compute method here. You just have to override the `name_get()` function, and the rest will be automatic.

You do need to set it as stored and indexed, though:

```suggestion
        store=True, index=True
```

yajo commented on project_sequence/models/project_project.py:
```diff
@@ -0,0 +1,45 @@
+            rec.name = rec.codesequence
+
+    @api.depends("name", "codesequence")
+    def _compute_display_name_project(self):
```
As explained above, you need to override the `name_get` function instead.

yajo reviewed (changes requested):
This looks quite good. I found some improvements though, could you take a look please? Thanks! 😊

rafaelbn reviewed (commented):
Dependency with `sequence_reset_period` not needed as base module has the functionality:

![imagen](https://user-images.githubusercontent.com/8736623/233404880-0423d619-2ae6-478d-9bb4-4dfe2ba7f3a8.png)

anddago78 commented:
Hi @rafaelbn i have removed the dependency with sequence_reset_period

fcvalgar reviewed (approved):
LGTM, Thank you @anddago78

rafaelbn reviewed (dismissed):
Functionally misunderstanding. Explanation in private to modify functionality. Sorry 😢

yajo committed:
[FIX] project_sequence: make compatible with hr_timesheet

`hr_timesheet` creates an analytic account by default. The method it uses to create it expects a preexisting name. But since we're making name not required, we're breaking other module's expectations.

To fix this problem, now the name sync is done before writing or creating records, and not after.

To make sure the problem doesn't happen anymore, we keep the `NOT NULL` requirement on project names. We just do it with a manual SQL constraint. This extra protection ensures compatibility with any other modules that expect always a value on the name.

Any possibly misconfigured sequence could produce sequence duplicates. I also add protection against that.

In tests, the project sequence was wrongly reset to 11 only once. Turns out that it survives the test savepoint, so I now reset it before each test instead. Tests are more reliable now. Besides, I added some more.

@moduon MT-1506

wip

wip

yajo committed:
[IMP] project_sequence: move sequence below name

When a project has a name, still the sequence is a very important field to be displayed. I move it below the project name.

@moduon MT-1506

yajo commented:
If runboat fails, you have to uninstall `project_wbs`. It introduces some incompatibilities. I have recorded that incompatibility in the manifest, so you can't install both and the problem is avoided.

yajo committed:
[FIX] project_wbs: make compatible with project_sequence

Support the use case of a `False` name coming in, which can happen when installing `project_sequence`.

@moduon MT-1506

yajo commented:
Due to https://github.com/sbidoul/runboat/issues/36, introducing the incompatibility with `project_wbs` made runboat die.

Now, I've fixed `project_wbs` so that it's resilient to the installation of `project_sequence`. Runboat should give no more problems and CI is ✅ now.

fcvalgar reviewed (approved):
LGTM, thank you @yajo

rafaelbn reviewed (commented):
Commented with @fcvalgar , display name standard without sequence

yajo committed:
[FIX] project_sequence: support preexisting sequenceless projects

These projects shouldn't display "False - Project name" in their display names.

@moduon MT-1506

rafaelbn reviewed (approved):
👍🏼 Thanks

An improvement por generalización will come before migration to 15

rafaelbn commented:
/ocabot merge patch
